### ISSUE TITLE
DNS:  $3BotDoubleName.grid.tf goes to the right 3bot and jimber terminates the ssl

### ISSUE THREAD

**Issue by robvanmieghem** [threefoldtech/home#114]

**Comment by xmonader** [comment]
AFAIK traefik template is deployed as 1 cluster with roundrobin and ALL sites are registered on each traefik instance

![3bot no HA](https://docs.google.com/drawings/d/e/2PACX-1vSFYG7Dyq7WVVocmtZpyZijBdI100cP2DBcIvFZT41R62mQicRIOiYllF9-e9tth_vj8IDJl_viMiuy/pub?w=2104&h=1017)

![3bot infrastructure](https://docs.google.com/drawings/d/e/2PACX-1vTMtU3bhf36SrGF6LoVZ_YELPrPi5Bih-LhIWsqpcUGsml1V4X4QBl-n8lf2fWRHFaq5nhAFDCP1P2v/pub?w=2104&h=1017)

also isn't the SSL termination is done by traefik not the 3bot?

**Comment by GlenDC** [comment]
The dns service should be connected to the tfchain network, this way it can update the 3bot records as blocks get applied and reverted. Easiest would be if you write it in Go, as you can than just use `tfchain` (Go) as a library, to have it as an embedded node to which you add your custom (ConsensusSetSubscriber-) module that would only focus on 3Bot transactions to apply/revert in your (etcd?) registry.

**Comment by zaibon** [comment]
> In a 3bot record on the blockchain, an ip address and doublename (xxxxx.yyyyy) are registered.

Just saw this sentence, and this actually raise a question.  There is nothing that force a threebot record to register an ip address. So it seems we also need to talk to another component then the blokchain to get the full flow.

So the component that will populate the configuration of the dns server needs to talk to both the blockchain to get the threebot double name (used as dns domain) and to the component that deploys the actual 3bot server on the grid cause he is the only one to know what is the ip of the 3bot server.

**Comment by GlenDC** [comment]
> Just saw this sentence, and this actually raise a question. There is nothing that force a threebot record to register an ip address.

The blockchain code related to the 3Bot should work in function of the expected data/relationships/behaviour. Thus if a 3Bot without an address (whether it is a hostname, IPv4 or Ipv6) does not make any sense, than it is easy enough for us to add a rule to enforce that all 3Bot records have at least 1 address at all times.

For now a name is also optional, and thus again if a 3Bot without a name does not make sense, please do tell me, and than we need to make sure that is not possible either.

Lastly if we should only accept 2 groups per 3bot name, than we need to align on what exactly the validation of a 3Bot name needs to be, so that it is in line with what is expected.

It would be rediculous that you need to talk to many different components, while all information (3Bot ID, name(s), address(es) and public key) should already be available in a single record.

Just align, and make sure we know what we want, and than the rest can be done, I'm sure of that.

**Comment by zaibon** [comment]
@robvanmieghem what are your thoughts on this ?

**Comment by zaibon** [comment]
I compile a document about this question: https://docs.google.com/document/d/1peUfAGxJE90lmNSWwOfBAuIZDEHoOYz0VnbYHZn2h2k/edit

As far as I can see this issue should be closed cause deprecated now.

@robvanmieghem please do so if agree.

**Comment by zaibon** [comment]
After a chat with @robvanmieghem, for the first version of the DNS infrastructure, we are going to implement a CoreDNS pluging that reads the 3bot double name from the blockchain using the 3 explorers available.

**Comment by xmonader** [comment]
- Traefik added tcp support in V 2.0 Alpha https://github.com/containous/traefik/tree/v2.0.0-alpha1 and It only supports Load balancing type https://docs.traefik.io/v2.0/routing/services/#general_1  I believe than can be worked around it if we give every service a unique label and only one instance.

**Comment by xmonader** [comment]
I did some experiments with the V2.0 Alpha

```toml
[entryPoints]
  [entryPoints.tcp]
  address = "localhost:8040"

[providers.file]

[tcp.services]
  [tcp.services.my-service.loadbalancer]
     [[tcp.services.my-service.loadbalancer.servers]]
        address = "127.0.0.1:9092"

[tcp.routers]
   [tcp.routers.to-my-service]
      rule = "HostSNI(`*`)"
      service = "my-service"
      [tcp.routers.to-my-service.tls]
         passthrough=true
```

Server:
```python
from gevent import monkey; monkey.patch_all()
import logging
from gevent.server import StreamServer


logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class Receiver(object):
    """ Interface for a receiver - mimics Twisted's protocols
    """
    def __init__(self):
        self.socket = None
        self.address = None

    def connection_made(self, socket, address):
        self.socket = socket
        self.address = address

    def connection_lost(self):
        pass

    def line_received(self, line):
        pass

    def send_line(self, line):
        self.socket.sendall(line + b'\n')


class EchoReceiver(Receiver):
    """ A basic implementation of a receiver which echoes back every line it
    receives.
    """
    def line_received(self, line):
        self.send_line(line)


def Handler(receiver_class):
    """ A basic connection handler that applies a receiver object to each
    connection.
    """
    def handle(socket, address):
        logger.info('Client (%s) connected', address)

        receiver = receiver_class()
        receiver.connection_made(socket, address)

        try:
            f = socket.makefile()

            while True:
                line = f.readline().strip()
                if line == "":
                    break
                logger.info('Received line from client: %s', line)
                receiver.line_received(line.encode())
            logger.info('Client (%s) disconnected.', address)

        except Exception as e:
            logger.exception(e)
        finally:
            try:
                f.close()
                receiver.connection_lost()
            except:
                pass
    return handle


server = StreamServer(('0.0.0.0', 9092), Handler(EchoReceiver), keyfile='server.key', certfile='server.crt')
logger.info('Server running')
server.serve_forever()
``` 

client:
```python
import time
from gevent import ssl, socket

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

# Require a certificate from the server. We used a self-signed certificate
# so here ca_certs must be the server certificate itself.
ssl_sock = ssl.wrap_socket(s,
                           ca_certs="server.crt",
                           cert_reqs=ssl.CERT_REQUIRED)

# ssl_sock.connect(('localhost', 9092))
ssl_sock.connect(('localhost', 8040))

ssl_sock.sendall(b'login superadmin password\n')

while True:
    time.sleep(5)
    ssl_sock.sendall(b'ping null\n')
    print(ssl_sock.recv(4096))

ssl_sock.close()

```

server.key and server.crt generated using
```
openssl req \
       -newkey rsa:2048 -nodes -keyout server.key \
       -x509 -days 365 -out server.crt
```

**Comment by zaibon** [comment]
> Traefik added tcp support in V 2.0 

@xmonader why exactly do we need TCP proxy in the first place ?

**Comment by xmonader** [comment]
@zaibon if the TLS termination is done by the 3bot application and not by traefik then tcp proxy is needed no?

**Comment by zaibon** [comment]
Yeah indeed, I forgot that 3bot uses plain TCP to communicate.
But now that I'm thinking about it. I think SSL is not the right technology to use in our case. SSL rely on certificates and certificate means you need an central authority that verify your identity. This really doesn't fit well in a fully decentralized world.

There are plenty of handshake protocol that allow 2 unknown peer to authenticate each other and create a secure channel. 
Examples : 
- https://ssbc.github.io/scuttlebutt-protocol-guide/index.html#handshake
- [Secret Handshake key exchange](https://dominictarr.github.io/secret-handshake-paper/shs.pdf)

These match way more with what we are trying to use and they have use the same kind of key that the 3bot already uses (ed25519)

So to me, as long as the DNS is in place and we can map a 3bot name to an IP, that's all we need. The rest of the communication can just start with an handshake which create a secure channel and we're good to go.
This handshake is also already implemented in different language and have been verified. Seems to me this is the shortest path to success.

**Comment by robvanmieghem** [comment]
You are right for machine to machine communication in which case you can use the 3bot's public key for the handshake  but in the case of a user using a webbrowser to surf to its 3bot, we do need regular https being terminated on the 3bot.

**Comment by zaibon** [comment]
Ok you're right, seems my comment doesn't really apply in this issue then.

I'm still don't fully understand how the 3bot will receive their certificate though

**Comment by robvanmieghem** [comment]
>I'm still don't fully understand how the 3bot will receive their certificate though

regular let's encrypt?
